Add Smcsrind and Sscsrind CSR YAML Files

syedowaisalishah opened the pull request:
This PR adds:
- Extension YAML files for `Smcsrind` and `Sscsrind`
- CSR YAMLs for:
  - MISELECT, MIREG to MIREG6
  - SISELECT, SIREG to SIREG6
  - VSISELECT, VSIREG to VSIREG6
  
  Closes #557

AFOliveira commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,30 @@
+length: MXLEN
+definedBy: Smcsrind
+description: |
+  description: |
```
This is duplicated

syedowaisalishah replied:
Thank you for catching that duplication, @AFOliveira! I've removed the redundant description: line from mireg.yaml and will double-check the others for similar issues before pushing the update.

dhower-qc commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,30 @@
+    location_rv64: 63-0
+    type: RW
+    description: Indirect register data selected by `miselect` value.
+    reset_value(): return 0x0;
```
all the reset values should be:

```yaml
    reset_value: UNDEFINED_LEGAL
```

dhower-qc commented:
Take a look at #594 

With that patch, you should be able to write the sw_read/sw_write functions for the indirect CSRs

syedowaisalishah commented:
Thanks for the clarification, @dhower-qc. I've updated all relevant CSR fields to use:
reset_value: UNDEFINED_LEGAL
I’ll also review the changes introduced in #594 and implement the sw_read()/sw_write() functions accordingly for the indirect CSRs. I’ll push the fixes shortly and re-request review once done.

ThinkOpenly commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,30 @@
+---
```
I don't see a line like that in any of the other files in `arch/csr`. I presume it's not needed, and should probably be removed.

dhower-qc replied:
FYI, the '---' is a "document" delimiter. It is indeed optional if you only have one document in the YAML file and no directives, which is the case here.

ThinkOpenly commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,30 @@
+$schema: csr_schema.json#
+kind: csr
+name: mireg
+long_name: Machine Indirect Register Alias 1
```
I certainly understand why the "1" was added, but let's hew to the ratified text in the spec, which does not have the "1".

ThinkOpenly commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,30 @@
+length: MXLEN
+definedBy: Smcsrind
+description: |
+  The `mireg` CSR is the primary machine-level indirect alias register used to access
+  register state as determined by the index held in `miselect`. This CSR is typically used
+  to read or write the full 32-bit or 64-bit value of the indirectly addressed register.
+
+  For 64-bit registers in RV32 systems, `mireg` is generally used to access the lower 32 bits.
+  Combined with higher-indexed aliases (e.g., `mireg4`), this provides full access to 64-bit values
+  across multiple CSRs.
+
+  If the value in `miselect` is not implemented, access is UNSPECIFIED. Implementations
+  are encouraged to raise an illegal instruction exception in such cases.
```
Where did this text come from?

ThinkOpenly commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,30 @@
+    location_rv32: 31-0
+    location_rv64: 63-0
+    type: RW
+    description: Indirect register data selected by `miselect` value.
```
Register state data of CSR selected by `miselect` value. Here, and elsewhere.

ThinkOpenly commented on arch/ext/Smcsrind.yaml:
```diff
@@ -0,0 +1,37 @@
+name: Smcsrind
+long_name: Machine Indirect CSR Access (Smcsrind)
+description: |
+  The `Smcsrind` extension defines indirect access to CSRs, allowing registers to be accessed through
+  an index value (`miselect`) rather than using a fixed CSR number. This reduces the need for allocating
+  large chunks of CSR address space, making it more scalable and flexible.
+
+  `Smcsrind` enables accessing an array of CSRs by index through dedicated `miselect` and `mireg*` registers
+  for machine-level access. The supervisor-level counterpart, `Sscsrind`, provides similar functionality but
+  without machine-level CSRs.
+
+  This extension is beneficial for accessing a set of registers that are not statically defined in the CSR space
+  and enables more flexible register allocation.
+
+  The machine-level registers are accessed through `miselect` and `mireg*` registers, while supervisor-level
+  registers are accessed through `siselect` and `sireg*` registers.
```
Should we use the text from section 5.1 "Introduction" here, edited minimally to make sense in this context?

ThinkOpenly commented on arch/ext/Smcsrind.yaml:
```diff
@@ -0,0 +1,37 @@
+      allOf:
+        - name: S
+          version: ~> 1.13
+    changes:
+      - Introduced indirect CSR access using `miselect`, `mireg*` for machine-level, and `siselect`, `sireg*` for supervisor-level CSRs.
+      - Defines the address spaces for indirect access in both machine and supervisor levels.
+      - Supports custom extension registers for each mode via indirect access
```
I'm thinking we don't need a "changes" stanza for extensions where only a single version exists, so this can be removed.

ThinkOpenly commented on arch/ext/Sscsrind.yaml:
```diff
@@ -0,0 +1,38 @@
+          version: ~> 1.13
+        - name: Smcsrind
+          version: ~> 1.0
+    changes:
+      - Introduced indirect CSR access for Supervisor-level using `siselect` and `sireg*` registers.
+      - Enables access to arrays of registers in Supervisor mode, making it more efficient to handle large numbers of registers.
+      - Supports custom extensions with indirect access for supervisor-level registers.
```
Similar to above, remove.

syedowaisalishah commented:
@ThinkOpenly Thank you for your detailed feedback! I've made the following updates based on your suggestions:
- Removed unnecessary --- lines.
- Adjusted the mireg name to match the ratified spec without the "1".
- Updated descriptions to reflect the correct text from the spec, including replacing the description with "Register state data of CSR selected by miselect value".
- Removed unnecessary "implies" and "changes" sections from the extension YAML files.
Let me know if anything else needs adjustment. Thanks again!

ThinkOpenly commented on arch/csr/Smcsrind/siselect.yaml:
```diff
@@ -10,24 +9,23 @@ length: SXLEN
-  exception in such cases to support emulation or error detection. Implementations
-  may make additional behaviors subject to platform or profile specifications.
+  exception in such cases to support emulation or error detection. Platforms or profiles
+  may define additional behaviors.
```
Where does this content come from?

ThinkOpenly commented on arch/csr/Smcsrind/siselect.yaml:
```diff
@@ -10,24 +9,23 @@ length: SXLEN
+  may define additional behaviors.
 
 fields:
   Select Value:
```
Can we have a field name with a space in it? I suggest "VALUE", like the previous case.

ThinkOpenly commented on arch/csr/Smcsrind/siselect.yaml:
```diff
@@ -36,6 +34,6 @@ fields:
     description: |
       Index value selecting the target register accessed through `sireg*`.
-      Acts as an indirect register selector. Values are extension-defined.
+      It serves as an indirect register selector. Values are defined by the extension.
```
I suggest changing this to (from the spec, modified a bit):
> Value ranges are allocated to dependent extensions, which specify the
> register state accessible via each `sireg`i register, for each `siselect` value.

ThinkOpenly commented on arch/csr/Smcsrind/vsireg2.yaml:
```diff
@@ -8,21 +7,24 @@ address: 0x252
+
+  The access behavior for each `vsiselect` + `vsireg2` pair is defined by the
+  extension that owns the `vsiselect` value. Ordinarily, the alias register
+  either accesses architectural state, reads zero, or raises an exception.
```
Where did this content come from?

ThinkOpenly commented on arch/ext/Smcsrind.yaml:
```diff
@@ -6,32 +6,17 @@ name: Smcsrind
+  an index value (`miselect`) instead of using fixed CSR numbers. This mechanism reduces the need for
+  large CSR address space allocation, making it more scalable and flexible. It enables the access of a set
+  of registers that are not statically defined in the CSR space. Machine-level registers are accessed
+  through `miselect` and `mireg*`, while supervisor-level registers are accessed through `siselect` and `sireg*`.
```
Where did this content come from?

syedowaisalishah commented:
@ThinkOpenly I've reviewed all the feedback and will revise the descriptions across mireg.yaml, siselect.yaml, vsireg.yaml, and other files to match the exact spec language where applicable. I'll also update the field names and fix the definedBy tags as per your suggestions. Thanks again for the thorough review — I’ll push the updated commits shortly.

Sorry for the mistake and any confusion caused.

syedowaisalishah commented on arch/csr/Smcsrind/vsireg.yaml:
```diff
@@ -8,7 +7,7 @@ address: 0x251
 priv_mode: VS
 length: VSXLEN
-definedBy: Sscsrind
+definedBy: Smcsrind
```
Correct me if I’m wrong, but you’re correct that the CSRIND bit in mstateen0 controls access to alias registers like vsireg and others, as per the spec. However, the vsireg CSR itself is defined by the Smcsrind extension, which is why it’s listed under definedBy: Smcsrind.

The critical point here is that while vsireg is part of the Smcsrind extension, its access is controlled indirectly via the vsiselect CSR, which in turn is influenced by the CSRIND bit from both the Sscsrind and Smcsrind extensions. So, while the ownership of the vsireg CSR lies with Smcsrind, the access control mechanism is shared, as outlined in the spec.

ThinkOpenly replied:
It's difficult to find definitive information. I found this:
> The machine-level extension Smcsrind encompasses all added CSRs and all behavior modifications for a hart, over all privilege levels. For a supervisor-level environment, extension Sscsrind is essentially the same as Smcsrind except excluding the machine-level CSRs and behavior not directly visible to supervisor level.

Given that, I agree with you that `vsireg` is defined by "Smcsrind".

syedowaisalishah commented:
@ThinkOpenly I’ve addressed all the suggestions you mentioned and tried my best to revise the descriptions in the YAML to closely match the exact wording from the spec. Let me know if I’ve missed anything.
Sorry for the oversight earlier!

ThinkOpenly commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,42 @@
+    location_rv32: 31-0
+    location_rv64: 63-0
+    type: RW
+    description: Register state of the CSR selected by the current `miselect` value.
```
Could you make all of the descriptions for the `mireg*` CSRs the same?
Suggest "Register state of the CSR selected by the current `miselect` value." (So, keep this one and change the others.)

syedowaisalishah replied:
Sure, I will do it!

ThinkOpenly commented on arch/ext/Smcsrind.yaml:
```diff
@@ -0,0 +1,44 @@
+    state: ratified
+    ratification_date: "2024-11"
+    url: "https://github.com/riscv/riscv-isa-manual/releases/tag/riscv-isa-release-482805d-2025-03-25"
+    requires:
+      allOf:
+        - name: S
+          version: ~> 1.13
```
"allOf" is superfluous here, since there is only one extension, "S". Suggest:
```
  requires: { name: S, version: "~> 1.13" }
```

AFOliveira reviewed (commented):
@syedowaisalishah There are several comments on this PR that likely have already been solved by recent changes, can you either answer them or resolve them so reviewers can know exactly what to look like when taking a second look at this PR? Also, Github will request that all of them are resolved when we want to merge.

syedowaisalishah commented:
@AFOliveira Thanks for the reminder! I'll go through all the comments, respond where needed, and resolve the ones that have been addressed by the recent changes. I'll make sure everything is clear for the reviewers before we proceed with the next review.

dhower-qc commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,42 @@
+    location_rv64: 63-0
+    type: RW
+    description: Register state of the CSR selected by the current `miselect` value
+    reset_value: UNDEFINED_LEGAL
```
needs a sw_write(csr_value) function like:

```yaml
  sw_write(csr_value): |
    Csr handle = indirect_csr_lookup(CSR[miselect].VALUE, 1);
    if (!handle.valid) {
      unimplemented_csr($encoding);
    }
    if (!handle.writable) {
      raise (ExceptionCode::IllegalInstruction, mode(), $encoding);
    }
    csr_sw_write(handle, csr_value.VALUE);
    return csr_hw_read(handle);
```

the rest of the mireg* registers will be similar, but with the last argument of indirect_csr_lookup adjusted to [2,3,4,5,6].

{v}sireg* will be similar, but will also need to check handle.mode (a PrivilegeMode type) and raise an IllegalInstruction if there is insufficient privilege.

dhower-qc reviewed (changes requested):
The comments generally apply to all the files.

syedowaisalishah commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,42 @@
+priv_mode: M
+length: MXLEN
+definedBy: Smcsrind
+description: |
+  The mireg machine indirect alias CSR is used to access another CSR's state
+  indirectly upon a read or write, as determined by the value of miselect.
+
+  The behavior upon accessing mireg from M-mode, while miselect holds a value
+  that is not implemented, is UNSPECIFIED.
+
+  [Note]
+  It is expected that implementations will typically raise an illegal instruction exception for
+  such accesses, so that, for example, they can be identified as software bugs. Platform
+  specs, profile specs, and/or the Privileged ISA spec may place more restrictions on
+  behavior for such accesses.
+
+  Attempts to access mireg while miselect holds a number in an allocated and implemented range
+  results in a specific behavior that, for each combination of miselect and mireg, is defined by the
+  extension to which the miselect value is allocated.
+
+  [Note]
+  Ordinarily, mireg will access register state, access read-only 0 state, or raise an
+  illegal instruction exception.
+
+  [Note]
+  For RV32, if an extension defines an indirectly accessed register as 64 bits wide, it is
+  recommended that the lower 32 bits of the register are accessed through mireg,
+  while the upper 32 bits are accessed through mireg4.
```
Understood, I’ll convert the description to the structured format as you suggested, including the long_name for the VALUE field. I’ll update the description accordingly.

ThinkOpenly commented:
@dhower-qc, is there a way to incorporate this:
> The behavior upon accessing sireg* from M-mode or S-mode, while siselect holds a value that is not implemented at supervisor level, is UNSPECIFIED.
> 
> It is recommended that implementations raise an illegal instruction exception for such accesses, to facilitate possible emulation (by M-mode) of these accesses.

Given the "recommended" term, do we need an extension parameter?

dhower-qc commented:
> @dhower-qc, is there a way to incorporate this:
> 
> > The behavior upon accessing sireg* from M-mode or S-mode, while siselect holds a value that is not implemented at supervisor level, is UNSPECIFIED.
> > It is recommended that implementations raise an illegal instruction exception for such accesses, to facilitate possible emulation (by M-mode) of these accesses.
> 
> Given the "recommended" term, do we need an extension parameter?

This is already handled as long as we call `unimplemented_csr` in IDL, since it's the same situation with direct CSR accesses. There is a `TRAP_ON_UNIMPLEMENTED_CSR` parameter, that does the following:

```
function unimplemented_csr {
  arguments Bits<INSTR_ENC_SIZE> encoding
  description {
    Either raises an IllegalInstruction exception or enters unpredictable state,
    depending on the setting of the TRAP_ON_UNIMPLEMENTED_CSR parameter.
  }
  body {
    if (TRAP_ON_UNIMPLEMENTED_CSR) {
      raise(ExceptionCode::IllegalInstruction, mode(), encoding);
    } else {
      unpredictable("Accessing an unimplmented CSR");
    }
  }
}
```

syedowaisalishah commented:
@dhower-qc  I’ve applied all the requested updates across mireg[1–6].yaml, sireg[1–6].yaml, and vsireg[1–6].yaml:

- Converted description fields to the structured format.

- Added long_name to the VALUE field.
 
- Implemented sw_read() and sw_write() methods, using the appropriate index in indirect_csr_lookup() for each file.

- Corrected the description format for {m,v,s}select as well.

dhower-qc commented on arch/csr/Smcsrind/mireg.yaml:
```diff
@@ -0,0 +1,71 @@
+        text: Register state of the CSR selected by the current `miselect` value
+    reset_value: UNDEFINED_LEGAL
+    sw_write(csr_value): |
+      Csr handle = indirect_csr_lookup(CSR[miselect].VALUE, 1);
+      if (!handle.valid) {
+      unimplemented_csr($encoding);
+      }
+      if (!handle.writable) {
+      raise (ExceptionCode::IllegalInstruction, mode(), $encoding);
+      }
+      csr_sw_write(handle, csr_value.VALUE);
+      return csr_hw_read(handle);
+sw_read(): |
```
Make sure the if bodies are indented, please.

syedowaisalishah replied:
sure i will do.

dhower-qc reviewed (changes requested):
Other than the small indentation request, looking good!!

dhower-qc reviewed (approved):
Looks good, assuming the regressions pass after pulling main.

Thanks for all the work!
